File: server/models/Wishlist.js
const mongoose = require('mongoose');

const { Schema, model } = mongoose;

const WishlistItemSchema = new Schema({
  product: { type: Schema.Types.ObjectId, ref: 'Product', required: true },
  snapshot: { type: Schema.Types.Mixed, default: null },
  addedAt: { type: Date, default: Date.now },
});

const WishlistSchema = new Schema({
  user: { type: Schema.Types.ObjectId, ref: 'User', sparse: true, index: true },
  guestId: { type: String, sparse: true, index: true }, // UUID for guest tracking
  items: { type: [WishlistItemSchema], default: [] },
  updatedAt: { type: Date, default: Date.now },
}, { timestamps: true });

module.exports = model('Wishlist', WishlistSchema);

File: server/models/wishlist.py
from pydantic import BaseModel, Field
from typing import List, Optional, Union
from datetime import datetime

class WishlistItem(BaseModel):
    product: str  # Reference to Product ID
    snapshot: Optional[dict] = None
    added_at: Optional[datetime] = None

    class Config:
        schema_extra = {
            "example": {
                "product": "product_id",
                "snapshot": {"name": "Product Name", "price": 100.0},
                "added_at": "2026-03-10T12:00:00Z"
            }
        }

class Wishlist(BaseModel):
    id: Optional[str] = Field(None, alias="_id")
    user: Optional[str] = None  # Reference to User ID
    guest_id: Optional[str] = None  # UUID for guest tracking
    items: List[WishlistItem] = []
    updated_at: Optional[datetime] = None

    class Config:
        allow_population_by_field_name = True
        schema_extra = {
            "example": {
                "user": "user_id",
                "guest_id": "guest123",
                "items": [
                    {
                        "product": "product_id",
                        "snapshot": {"name": "Product Name", "price": 100.0},
                        "added_at": "2026-03-10T12:00:00Z"
                    }
                ],
                "updated_at": "2026-03-10T12:00:00Z"
            }
        }